File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/Fingerprint.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface FingerprintLib: Library {
    fun Fingerprint_destroy(handle: Pointer)
    fun Fingerprint_compute(schematic: Pointer, preset: Slice, write: Pointer): ResultUnitInt
    fun Fingerprint_signature_json(schematic: Pointer, preset: Slice, write: Pointer): ResultUnitInt
    fun Fingerprint_footprint_distance(a: Pointer, b: Pointer, preset: Slice): ResultFloatInt
    fun Fingerprint_footprint_json(schematic: Pointer, preset: Slice, write: Pointer): ResultUnitInt
    fun Fingerprint_is_duplicate(a: Pointer, b: Pointer, preset: Slice): ResultByteInt
}
/** Namespace type for the fingerprint free functions (PORTING rule 12).
*/
class Fingerprint internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class FingerprintCleaner(val handle: Pointer, val lib: FingerprintLib) : Runnable {
        override fun run() {
            lib.Fingerprint_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, Fingerprint.FingerprintCleaner(handle, Fingerprint.lib));
    }

    companion object {
        internal val libClass: Class<FingerprintLib> = FingerprintLib::class.java
        internal val lib: FingerprintLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** The fingerprint of a schematic for the given preset, as a hex string.
        *Errors with `InvalidArgument` on an unknown preset.
        */
        fun compute(schematic: Schematic, preset: String): Result<String> {
            val presetSliceMemory = PrimitiveArrayTools.borrowUtf8(preset)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.Fingerprint_compute(schematic.handle, presetSliceMemory.slice, write);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    
                    val returnString = DW.writeToString(write)
                    return returnString.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                presetSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** The structural signature (JSON) of a schematic for the given preset.
        */
        fun signatureJson(schematic: Schematic, preset: String): Result<String> {
            val presetSliceMemory = PrimitiveArrayTools.borrowUtf8(preset)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.Fingerprint_signature_json(schematic.handle, presetSliceMemory.slice, write);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    
                    val returnString = DW.writeToString(write)
                    return returnString.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                presetSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Translation-invariant fuzzy distance between two builds' footprints.
        */
        fun footprintDistance(a: Schematic, b: Schematic, preset: String): Result<Float> {
            val presetSliceMemory = PrimitiveArrayTools.borrowUtf8(preset)
            
            val returnVal = lib.Fingerprint_footprint_distance(a.handle, b.handle, presetSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    return (nativeOkVal).ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                presetSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** The schematic's translation/scale-invariant FFT shape footprint as a
        *JSON array of floats.
        */
        fun footprintJson(schematic: Schematic, preset: String): Result<String> {
            val presetSliceMemory = PrimitiveArrayTools.borrowUtf8(preset)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.Fingerprint_footprint_json(schematic.handle, presetSliceMemory.slice, write);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    
                    val returnString = DW.writeToString(write)
                    return returnString.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                presetSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Whether two schematics share the same fingerprint for the given preset.
        */
        fun isDuplicate(a: Schematic, b: Schematic, preset: String): Result<Boolean> {
            val presetSliceMemory = PrimitiveArrayTools.borrowUtf8(preset)
            
            val returnVal = lib.Fingerprint_is_duplicate(a.handle, b.handle, presetSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    return (nativeOkVal > 0).ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                presetSliceMemory.close()
            }
        }
    }

}

File: bindings/php/Fingerprint.php
<?php
namespace Stencil;

final class Fingerprint {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned) {
        $this->ptr = $ptr;
        $this->owned = $owned;
    }

    public static function compute( $schematic, string $preset) {
        $__n1 = strlen($preset);
        $__view1 = Lib::ffi()->new('DiplomatStringView');
        if ($__n1 > 0) {
            $__buf1 = Lib::ffi()->new("uint8_t[" . $__n1 . "]", false);
            \FFI::memcpy($__buf1, $preset, $__n1);
            $__view1->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf1[0]));
        } else {
            $__view1->data = null;
        }
        $__view1->len = $__n1;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Fingerprint_compute($schematic->ptr, $__view1, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function signatureJson( $schematic, string $preset) {
        $__n1 = strlen($preset);
        $__view1 = Lib::ffi()->new('DiplomatStringView');
        if ($__n1 > 0) {
            $__buf1 = Lib::ffi()->new("uint8_t[" . $__n1 . "]", false);
            \FFI::memcpy($__buf1, $preset, $__n1);
            $__view1->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf1[0]));
        } else {
            $__view1->data = null;
        }
        $__view1->len = $__n1;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Fingerprint_signature_json($schematic->ptr, $__view1, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function footprintDistance( $a,  $b, string $preset) {
        $__n2 = strlen($preset);
        $__view2 = Lib::ffi()->new('DiplomatStringView');
        if ($__n2 > 0) {
            $__buf2 = Lib::ffi()->new("uint8_t[" . $__n2 . "]", false);
            \FFI::memcpy($__buf2, $preset, $__n2);
            $__view2->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf2[0]));
        } else {
            $__view2->data = null;
        }
        $__view2->len = $__n2;
        $result = Lib::ffi()->Fingerprint_footprint_distance($a->ptr, $b->ptr, $__view2);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return $result->ok;
    }

    public static function footprintJson( $schematic, string $preset) {
        $__n1 = strlen($preset);
        $__view1 = Lib::ffi()->new('DiplomatStringView');
        if ($__n1 > 0) {
            $__buf1 = Lib::ffi()->new("uint8_t[" . $__n1 . "]", false);
            \FFI::memcpy($__buf1, $preset, $__n1);
            $__view1->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf1[0]));
        } else {
            $__view1->data = null;
        }
        $__view1->len = $__n1;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Fingerprint_footprint_json($schematic->ptr, $__view1, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function isDuplicate( $a,  $b, string $preset) {
        $__n2 = strlen($preset);
        $__view2 = Lib::ffi()->new('DiplomatStringView');
        if ($__n2 > 0) {
            $__buf2 = Lib::ffi()->new("uint8_t[" . $__n2 . "]", false);
            \FFI::memcpy($__buf2, $preset, $__n2);
            $__view2->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf2[0]));
        } else {
            $__view2->data = null;
        }
        $__view2->len = $__n2;
        $result = Lib::ffi()->Fingerprint_is_duplicate($a->ptr, $b->ptr, $__view2);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return $result->ok;
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->Fingerprint_destroy($this->ptr);
        }
    }
}
